transports/quic/src/transport.rs

```diff
@@ -0,0 +1,641 @@
+
+//! Implementation of the [`Transport`] trait for QUIC.
+//!
+//! Combines all the objects in the other modules to implement the trait.
+
+use crate::connection::Connection;
+use crate::endpoint::ToEndpoint;
+use crate::{endpoint::EndpointChannel, muxer::QuicMuxer, upgrade::Upgrade};
+use crate::{Config, ConnectionError};
+
+#[cfg(feature = "async-std")]
+mod async_std;
+#[cfg(feature = "tokio")]
+mod tokio;
+#[cfg(feature = "async-std")]
+pub use async_std::{AsyncStd, AsyncStdTransport};
+use futures::future::BoxFuture;
+#[cfg(feature = "tokio")]
+pub use tokio::{Tokio, TokioTransport};
+
+use futures::channel::{mpsc, oneshot};
+use futures::ready;
+use futures::stream::StreamExt;
+use futures::{prelude::*, stream::SelectAll};
+
+use if_watch::{IfEvent, IfWatcher};
+
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerId, TransportError as CoreTransportError, TransportEvent},
+    PeerId, Transport,
+};
+use std::collections::VecDeque;
+use std::io;
+use std::marker::PhantomData;
+use std::net::{IpAddr, UdpSocket};
+use std::task::Waker;
+use std::{
+    net::SocketAddr,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+#[derive(Debug)]
+pub struct QuicTransport<P> {
+    config: Config,
+    listeners: SelectAll<Listener>,
+    /// Dialer for Ipv4 addresses if no matching listener exists.
+    ipv4_dialer: Option<Dialer>,
+    /// Dialer for Ipv6 addresses if no matching listener exists.
+    ipv6_dialer: Option<Dialer>,
+
+    _marker: PhantomData<P>,
+}
+
+impl<P> QuicTransport<P> {
+    pub fn new(config: Config) -> Self {
+        Self {
+            listeners: SelectAll::new(),
+            config,
+            ipv4_dialer: None,
+            ipv6_dialer: None,
+            _marker: Default::default(),
+        }
+    }
+}
+
+/// Error that can happen on the transport.
+#[derive(Debug, thiserror::Error)]
+pub enum TransportError {
+    /// Error while trying to reach a remote.
+    #[error("{0}")]
+    Reach(quinn_proto::ConnectError),
+    /// Error after the remote has been reached.
+    #[error("{0}")]
+    Established(ConnectionError),
+
+    #[error("{0}")]
+    Io(#[from] std::io::Error),
+
+    /// The task driving the endpoint has crashed.
+    #[error("Endpoint driver crashed")]
+    EndpointDriverCrashed,
+}
+
+impl<P: Provider> Transport for QuicTransport<P> {
+    type Output = (PeerId, QuicMuxer);
+    type Error = TransportError;
+    type ListenerUpgrade = Upgrade;
+    type Dial = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn listen_on(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<ListenerId, CoreTransportError<Self::Error>> {
+        let socket_addr = multiaddr_to_socketaddr(&addr)
+            .ok_or(CoreTransportError::MultiaddrNotSupported(addr))?;
+        let listener_id = ListenerId::new();
+        let listener = Listener::new::<P>(listener_id, socket_addr, self.config.clone())
+            .map_err(CoreTransportError::Other)?;
```

Comment:
Can we implement `From` instead of calling `map_err` ourselves?

Reply:
Included in 6f284cc.